src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+        }
+
+        Ok(())
+    }
+
+    /// Notify this poller.
+    pub(super) fn notify(&self) -> io::Result<()> {
+        // Push the notify packet into the IOCP.
+        self.port.post(0, 0, self.notifier.clone())
+    }
+
+    /// Run an update on a packet.
+    fn update_packet(&self, mut packet: Packet) -> io::Result<()> {
+        loop {
+            // If we are currently polling, we need to update the packet immediately.
+            if self.polling.load(Ordering::Acquire) {
+                packet.update()?;
+                return Ok(());
+            }
+
+            // Try to queue the update.
+            match self.pending_updates.push(packet) {
+                Ok(()) => return Ok(()),
+                Err(p) => packet = p.into_inner(),
+            }
+
+            // If we failed to queue the update, we need to drain the queue first.
+            self.drain_update_queue(true)?;
+        }
+    }
+
+    /// Drain the update queue.
+    fn drain_update_queue(&self, limit: bool) -> io::Result<()> {
+        let max = if limit {
+            self.pending_updates.capacity().unwrap()
+        } else {
+            std::usize::MAX
+        };
+
+        // Only drain the queue's capacity, since this could in theory run forever.
+        for _ in 0..max {
+            if let Ok(packet) = self.pending_updates.pop() {
+                packet.update()?;
+            } else {
+                return Ok(());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Get a handle to the AFD reference.
+    fn afd_handle(&self) -> io::Result<Arc<Afd<Packet>>> {
+        const AFD_MAX_SIZE: usize = 32;
+
+        // See if there are any existing AFD instances that we can use.
+        let mut afd_handles = self.afd.lock().unwrap_or_else(|e| e.into_inner());
+        if let Some(handle) = afd_handles.iter().find(|h| {
+            let ref_count = Arc::strong_count(h).saturating_sub(1);
+            ref_count < AFD_MAX_SIZE
+        }) {
+            return Ok(handle.clone());
+        }
+
+        // Create a new AFD instance.
+        let afd = Arc::new(Afd::new()?);
+
+        // Register the AFD instance with the I/O completion port.
+        self.port.register(&*afd, true)?;
+
+        // Insert a copy of the AFD instance into the list.
+        afd_handles.push(afd.clone());
+
+        Ok(afd)
+    }
+}
+
+impl AsRawHandle for Poller {
+    fn as_raw_handle(&self) -> RawHandle {
+        self.port.as_raw_handle()
+    }
+}
+
+#[cfg(not(polling_no_io_safety))]
+impl AsHandle for Poller {
+    fn as_handle(&self) -> BorrowedHandle<'_> {
+        unsafe { BorrowedHandle::borrow_raw(self.as_raw_handle()) }
+    }
+}
+
+/// The container for events.
+pub(super) struct Events {
+    /// List of IOCP packets.
+    packets: Vec<Event>,
+}
+
+unsafe impl Send for Events {}
+
+impl Events {
+    /// Creates an empty list of events.
+    pub(super) fn new() -> Events {
```

Comment:
why isn't this `pub`?

Reply:
`Events` is internal to the `Poller` struct and used as an event container. That being said, I'd like to make the `Events` struct public so that we don't go through a `Vec<Event>` for events, see #66.